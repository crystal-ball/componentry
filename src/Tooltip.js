/* eslint-disable  react/prop-types */
import generateContainer from './toggleable/generate-container';
const Container = generateContainer('tooltip');

import { getTextWidth } from './utils/dom';

/**
 * The Tooltip component creates an expandable info container on hover.
 *
 * TODO:
 * - Check body overflow on render and adjust width if needed
 * @class Tooltip
 * @constructor
 * @extends React.Component
 */
export default class Tooltip extends Container {
  // Control arias used by subcomponents
  contentArias = { id: true, role: 'tooltip' }
  triggerArias = { describedby: true }

  /**
   * Internal cache for width of tooltip content. Set after calculating content
   * width and reused on subsequent renders if content text has not changed.
   */
  contentWidth = null;
  /**
   * Internal cache for tooltip content. Used to check if the content has changed
   * between showings of tooltip.
   */
  content = null;

  /**
   * Handle calling user hooks and toggling internal state if component is not
   * controlled.
   */
  toggleActive = e => {
    let { active } = this.props;
    // Handle uncontrolled tooltip with internal state
    if (active === undefined) { active = this.state.active; }

    if (!active) {
      // Tooltip is not active, so it is transitioning to shown
      this.props.onActivate(this, e);

      // Position absolute tooltip is constrained by the parent width. Set tooltip
      // width to content width to overflow parent bounds
      const contentElement = document.getElementById(this.guid);
      const content = contentElement.innerText;
      this.content = content;

      if (content === this.content && this.contentWidth) {
        // If width has already been calculated and content has not changed, use
        // cached width for performance
        contentElement.style.width = `${this.contentWidth}px`;
      } else {
        // Get all styles of content element, set width and cache
        const styles = window.getComputedStyle(contentElement);
        // Get padding, font size and font family of content
        const width = getTextWidth(content, `${styles.fontSize} ${styles.fontFamily}`)
          + parseFloat(styles.paddingLeft)
          + parseFloat(styles.paddingRight)
          + 1;

        contentElement.style.width = `${width}px`;
        this.contentWidth = width;
      }
    } else {
      // Tooltip is active, so it is transitioning to hidden
      this.props.onDeactivate(this, e);
    }

    if (this.props.active === undefined) {
      // State is not controlled, update internal state
      this.setState({ active: !active });
    }

    if (!active) {
      this.props.onActivated(this, e);
    } else {
      this.props.onDeactivated(this, e);
    }
  }
}
