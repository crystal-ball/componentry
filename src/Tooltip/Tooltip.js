import activeContainer from '../factories/active-container-component'
import activeContent from '../factories/active-content-component'
import activeTrigger from '../factories/active-trigger-component'

/**
 * [Tooltip component 📝](https://componentry.design/components/tooltip)
 */
const Tooltip = activeContainer('tooltip', {
  escEvents: true,
  mouseEvents: true,
})
export default Tooltip

/**
 * [Tooltip trigger component 📝](https://componentry.design/components/tooltip)
 */
Tooltip.Content = activeContent('tooltip', {
  arias: { id: true, role: 'tooltip', hidden: true },
  positioned: true,
})

/**
 * [Tooltip content component 📝](https://componentry.design/components/tooltip)
 */
Tooltip.Trigger = activeTrigger('tooltip', {
  arias: { describedby: true },
})
