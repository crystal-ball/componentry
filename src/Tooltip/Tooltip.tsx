import { activeContainerBuilder } from '../utils/active-container-component-builder'
import { activeActionBuilder } from '../utils/active-action-component-builder'
import { activeContentBuilder } from '../utils/active-content-component-builder'
import {
  ActiveActionBaseProps,
  ActiveContainerBaseProps,
  ActiveContentBaseProps,
  ComponentBaseProps,
} from '../utils/types'

export interface TooltipProps
  extends ActiveContainerBaseProps,
    ComponentBaseProps<'div'> {}

export interface TooltipActionProps
  extends ActiveActionBaseProps,
    ComponentBaseProps<'button'> {
  /** Display variant */
  variant?: 'primary'
}

export interface TooltipContentProps
  extends ActiveContentBaseProps,
    ComponentBaseProps<'div'> {
  /** Display variant */
  variant?: 'primary'
}

export interface Tooltip {
  (props: TooltipProps): React.ReactElement
  /**
   * [Tooltip action component üìù](https://componentry.design/components/tooltip)
   */
  Action: React.FC<TooltipActionProps>
  /**
   * [Tooltip content component üìù](https://componentry.design/components/tooltip)
   */
  Content: React.FC<TooltipContentProps>
}

/**
 * [Tooltip component üìù](https://componentry.design/components/tooltip)
 */
export const Tooltip = activeContainerBuilder<TooltipProps>('Tooltip', {
  escEvents: true,
  mouseEvents: true,
}) as Tooltip

Tooltip.Action = activeActionBuilder<TooltipActionProps>('TooltipAction', {
  aria: { describedby: true },
})

Tooltip.Content = activeContentBuilder<TooltipContentProps>('TooltipContent', {
  aria: { id: true, role: 'tooltip', hidden: true },
  positioned: true,
})
