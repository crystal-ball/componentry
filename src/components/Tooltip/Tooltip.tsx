import React from 'react'
import { activeActionBuilder } from '../../utils/active-action-component-builder'
import { activeContainerBuilder } from '../../utils/active-container-component-builder'
import { activeContentBuilder } from '../../utils/active-content-component-builder'
import {
  ActiveActionBaseProps,
  ActiveContainerBaseProps,
  ActiveContentBaseProps,
} from '../../utils/base-types'
import { UtilityProps } from '../../utils/utility-classes'
import { Link } from '../Link/Link'

export interface TooltipProps
  extends ActiveContainerBaseProps,
    UtilityProps,
    React.ComponentPropsWithoutRef<'div'> {}

export interface TooltipActionProps
  extends ActiveActionBaseProps,
    UtilityProps,
    React.ComponentPropsWithoutRef<'button'> {
  /** Display variant */
  variant?: 'primary'
}

export interface TooltipContentProps
  extends ActiveContentBaseProps,
    UtilityProps,
    React.ComponentPropsWithoutRef<'div'> {
  /** Display variant */
  variant?: 'primary'
}

export interface Tooltip {
  (props: TooltipProps): React.ReactElement
  /**
   * [Tooltip action component üìù](https://componentry.design/components/tooltip)
   */
  Action: React.FC<TooltipActionProps>
  /**
   * [Tooltip content component üìù](https://componentry.design/components/tooltip)
   */
  Content: React.FC<TooltipContentProps>
}

/**
 * [Tooltip component üìù](https://componentry.design/components/tooltip)
 * @experimental
 */
export const Tooltip = activeContainerBuilder('Tooltip', {
  escEvents: true,
  mouseEvents: true,
}) as Tooltip

Tooltip.Action = activeActionBuilder('TooltipAction', {
  aria: { describedby: true },
  defaultAs: Link,
})

Tooltip.Content = activeContentBuilder('TooltipContent', {
  aria: { id: true, role: 'tooltip', hidden: true },
  defaultAs: TooltipContentElement,
})

function TooltipContentElement({
  children,
  renderArrow = true,
  ...rest
}: {
  children: React.ReactNode
  renderArrow: boolean
}) {
  return (
    <div {...rest}>
      {renderArrow && <div className='C9Y-TooltipContentArrow' />}
      <div className='C9Y-TooltipContentContents'>{children}</div>
    </div>
  )
}
