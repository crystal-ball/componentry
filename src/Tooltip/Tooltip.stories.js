/* global module */

import React from 'react'
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'

import Tooltip from './Tooltip'

storiesOf('Tooltip', module).add('<Tooltip />', () => (
  <Tooltip
    onActivate={action('onActivate event')}
    onActivated={action('onActivated event')}
    onDeactivate={action('onDeactivate event')}
    onDeactivated={action('onDeactivated event')}
  >
    <Tooltip.Trigger>Toggle</Tooltip.Trigger>
    <Tooltip.Content>
      Only 8% of the worldâ€™s currency is physical money, the rest only exists on
      computers.
    </Tooltip.Content>
  </Tooltip>
))
