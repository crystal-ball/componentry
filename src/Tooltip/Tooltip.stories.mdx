import { Meta, Title, Props } from '@storybook/addon-docs/blocks'
import { action } from '@storybook/addon-actions'
import Tooltip from './Tooltip'

<Meta title='Components|Tooltip' component={Tooltip} />

# Tooltip

<Tooltip
  onActivate={action('onActivate event')}
  onActivated={action('onActivated event')}
  onDeactivate={action('onDeactivate event')}
  onDeactivated={action('onDeactivated event')}
>
  <Tooltip.Trigger>Fun fact</Tooltip.Trigger>
  <Tooltip.Content>
    Only 8% of the worldâ€™s currency is physical money, the rest only exists on computers.
  </Tooltip.Content>
</Tooltip>
