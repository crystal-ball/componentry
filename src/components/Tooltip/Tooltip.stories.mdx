import { Meta } from '@storybook/addon-docs'
import { BADGE } from '@geometricpanda/storybook-addon-badges'

import { Tooltip } from './Tooltip'

<Meta
  title='Components/Tooltip'
  component={Tooltip}
  parameters={{ badges: [BADGE.EXPERIMENTAL] }}
/>

# Tooltip

Tooltip extends Active with support for mouse events and styling to match typical tooltip
titles.

<Tooltip
  onActivate={console.log}
  onActivated={console.log}
  onDeactivate={console.log}
  onDeactivated={console.log}
>
  <Tooltip.Action>Fun fact</Tooltip.Action>
  <Tooltip.Content>
    Only 8% of the worldâ€™s currency is physical money, the rest only exists on computers.
  </Tooltip.Content>
</Tooltip>
