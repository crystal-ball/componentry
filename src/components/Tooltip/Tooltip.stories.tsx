import type { Meta, StoryObj } from '@storybook/react'

import { Tooltip } from './Tooltip'

const meta: Meta<typeof Tooltip> = {
  component: Tooltip,
}

export default meta
type Story = StoryObj<typeof Tooltip>

export const Primary: Story = {
  render: () => (
    <Tooltip active>
      <Tooltip.Action>Fun fact</Tooltip.Action>
      <Tooltip.Content>
        Only 8% of the world's currency is physical money, the rest only exists on
        computers.
      </Tooltip.Content>
    </Tooltip>
  ),
}
