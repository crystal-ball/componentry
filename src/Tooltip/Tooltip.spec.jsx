import React from 'react'
import renderer from 'react-test-renderer'

import Tooltip from './Tooltip'
// import activationTestSuite from '../utils-test/activation-tests'
import elementTests from '../utils-test/element-tests'
import setupActiveContext from '../utils-test/setup-active-context'

describe('<Tooltip />', () => {
  // Basic library activation test suite
  // TODO: Fix me
  // activationTestSuite(Popover)
  // Basic library element test suite
  elementTests(Tooltip)
  elementTests(Tooltip.Trigger, setupActiveContext())
  elementTests(Tooltip.Content, setupActiveContext())
})

// Snapshots
// ---------------------------------------------------------------------------
describe('<Drawer /> snapshots', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <Tooltip>
          <Tooltip.Trigger>Trigger</Tooltip.Trigger>
          <Tooltip.Content>Content</Tooltip.Content>
        </Tooltip>,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
