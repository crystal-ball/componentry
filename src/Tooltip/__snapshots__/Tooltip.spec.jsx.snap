// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="tooltip"
  data-id="guid"
  data-test="tooltip-container"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <button
    aria-describedby="guid"
    className="btn tooltip-toggle"
    data-test="tooltip-trigger"
    onClick={[Function]}
    type="button"
    value={undefined}
  >
    Trigger
  </button>
  <div
    className="tooltip-content-container"
  >
    <div
      aria-hidden="true"
      className="tooltip-content"
      data-test="tooltip-content"
      id="guid"
      role="tooltip"
    >
      <div
        className="tip-container"
      >
        <div
          className="tip"
        />
      </div>
      Content
    </div>
  </div>
</div>
`;
